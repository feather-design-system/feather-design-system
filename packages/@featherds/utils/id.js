const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
