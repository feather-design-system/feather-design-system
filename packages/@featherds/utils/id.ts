const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
