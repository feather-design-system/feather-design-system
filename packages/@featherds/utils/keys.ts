const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
