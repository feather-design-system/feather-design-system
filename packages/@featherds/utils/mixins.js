var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
