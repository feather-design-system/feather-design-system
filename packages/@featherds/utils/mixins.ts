const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
