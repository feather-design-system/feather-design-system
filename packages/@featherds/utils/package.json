{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
