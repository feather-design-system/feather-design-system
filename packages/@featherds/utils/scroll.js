var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
