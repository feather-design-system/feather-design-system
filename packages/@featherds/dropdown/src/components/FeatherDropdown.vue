<template>
  <FeatherMenu
    ref="menu"
    class="feather-dropdown-menu-container"
    data-ref-id="feather-dropdown-menu-container"
    :cover="cover"
    :open="localOpen"
    :right="right"
    @trigger-click="handleTriggerClick"
    @outside-click="handleOutsideClick"
    @close="handleClose"
  >
    <template v-slot:trigger>
      <slot name="trigger"></slot>
    </template>
    <template v-slot:default="slotProps">
      <ul
        @keydown="handleKeydown"
        :aria-labelledby="slotProps.labelId"
        class="feather-dropdown"
        :class="{ standard: standard }"
      >
        <slot></slot>
      </ul>
    </template>
  </FeatherMenu>
</template>
<script lang="ts">
import { ref, watch, nextTick, defineComponent } from "vue";
import { FeatherMenu } from "@featherds/menu";
import { Code } from "@featherds/utils/keys";
import { useDropdownService } from "../composables/DropdownService";
export const props = {
  modelValue: {
    type: Boolean,
    default: false,
  },
  cover: {
    type: Boolean,
    default: false,
  },
  right: {
    type: Boolean,
    default: false,
  },
  standard: {
    type: Boolean,
    default: false,
  },
} as const;
export const emits = {
  "update:modelValue": (_v: boolean) => true,
};
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits,
  props,
  setup(props, context) {
    const menu = ref(null);
    const localOpen = ref(props.modelValue);
    const dropdownService = useDropdownService();

    watch(
      () => props.modelValue,
      (v) => (localOpen.value = v)
    );
    watch(localOpen, (open) => {
      if (open) {
        nextTick(() => {
          const menuComponent = menu.value as unknown as { menu: HTMLElement };
          const items = Array.from(menuComponent.menu.querySelectorAll("a"));
          dropdownService.setItems(items);
          dropdownService.selectFirst();
        });
      } else {
        dropdownService.blur();
      }
      context.emit("update:modelValue", open);
    });
    const handleClose = () => {
      localOpen.value = false;
      const menuComponent = menu.value as unknown as { trigger: HTMLElement };
      if (menuComponent.trigger && menuComponent.trigger.focus) {
        menuComponent.trigger.focus();
      }
    };
    const handleTriggerClick = () => {
      localOpen.value = !localOpen.value;
    };
    const handleOutsideClick = () => {
      localOpen.value = false;
    };
    const handleKeydown = (e: KeyboardEvent) => {
      switch (e.code) {
        //next
        case Code.DOWN:
        case Code.RIGHT:
          dropdownService.selectNext();
          e.preventDefault();
          break;
        //previous
        case Code.UP:
        case Code.LEFT:
          dropdownService.selectPrevious();
          e.preventDefault();
          break;
        case Code.ESCAPE: //ESC
          handleClose();
          break;
        default:
          break;
      }
    };

    return {
      menu,
      dropdownService,
      localOpen,
      handleClose,
      handleTriggerClick,
      handleOutsideClick,
      handleKeydown,
    };
  },
  components: {
    FeatherMenu,
  },
});
</script>

<style lang="scss" scoped>
.feather-dropdown-menu-container {
  :deep([menu-trigger]) {
    margin-top: 0;
    margin-bottom: 0;
  }
}
.feather-dropdown {
  margin: 0;
  padding: 0;
}
.feather-dropdown.standard {
  width: 20rem;
}
.feather-dropdown {
  min-width: 7rem;
  overflow-y: auto;
  padding: 0.5rem 0;
  margin: 0;
}
</style>

<style lang="scss">
@import "../../scss/mixins";
.feather-dropdown {
  @include dropdown-menu-height(6);
}
</style>
