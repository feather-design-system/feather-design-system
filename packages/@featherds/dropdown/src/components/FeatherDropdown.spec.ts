import { defineComponent, nextTick } from "vue";
import { mount } from "@vue/test-utils";
import FeatherDropdown from "./FeatherDropdown.vue";
import axe from "@featherds/utils/test/axe";
import { getCalls } from "@featherds/utils/test/calls";
import { Code } from "@featherds/utils/keys";

import * as id from "@featherds/utils/id";

jest.spyOn(id, "getSafeId").mockImplementation((x) => x);
const getWrapper = (options: Record<string, unknown> = {}) => {
  return mount(FeatherDropdown, options);
};

const getTrigger = () => ({
  template: "<button menu-trigger>TEST</button>",
  methods: {
    focus: () => {},
  },
});
const getLi = (disabled = false) =>
  defineComponent({
    template: `<li>
    <a>TEST</a>
  </li>`,
    data() {
      return { disabled: disabled, focused: false };
    },
    computed: {
      _dropdownItem() {
        return true;
      },
    },
    methods: {
      focus() {
        this.focused = true;
      },
    },
  });

describe("FeatherDropdown.vue", () => {
  it("should pass through cover property", () => {
    const slots = {
      trigger: getTrigger(),
      default: [getLi(), getLi()],
    };
    const wrapper = getWrapper({
      slots,
      props: {
        cover: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should pass through right property", () => {
    const slots = {
      trigger: getTrigger(),
      default: [getLi(), getLi()],
    };
    const wrapper = getWrapper({
      slots,
      props: {
        right: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  describe("keydown", () => {
    describe("should call select next when", () => {
      it("down is pressed", async () => {
        const slots = {
          trigger: getTrigger(),
          default: [getLi(), getLi()],
        };
        const wrapper = getWrapper({ slots });
        jest.spyOn(wrapper.vm.dropdownService, "selectNext");
        await nextTick();
        await wrapper.setProps({
          modelValue: true,
        });
        await nextTick();
        const ul = wrapper.find("ul");
        await ul.trigger("keydown", { code: Code.DOWN });
        expect(wrapper.vm.dropdownService.selectNext).toHaveBeenCalled();
      });
      it("right is pressed", async () => {
        const slots = {
          trigger: getTrigger(),
          default: [getLi(), getLi()],
        };
        const wrapper = getWrapper({ slots });
        jest.spyOn(wrapper.vm.dropdownService, "selectNext");
        await nextTick();
        await wrapper.setProps({
          modelValue: true,
        });
        await nextTick();
        const ul = wrapper.find("ul");
        await ul.trigger("keydown", { code: Code.RIGHT });
        expect(wrapper.vm.dropdownService.selectNext).toHaveBeenCalled();
      });
    });
    describe("should call select previous when", () => {
      it("left is pressed", async () => {
        const slots = {
          trigger: getTrigger(),
          default: [getLi(), getLi()],
        };
        const wrapper = getWrapper({ slots });
        jest.spyOn(wrapper.vm.dropdownService, "selectPrevious");
        await nextTick();
        await wrapper.setProps({
          modelValue: true,
        });
        await nextTick();
        const ul = wrapper.find("ul");
        await ul.trigger("keydown", { code: Code.LEFT });
        expect(wrapper.vm.dropdownService.selectPrevious).toHaveBeenCalled();
      });
      it("up is pressed", async () => {
        const slots = {
          trigger: getTrigger(),
          default: [getLi(), getLi()],
        };
        const wrapper = getWrapper({ slots });
        jest.spyOn(wrapper.vm.dropdownService, "selectPrevious");
        await nextTick();
        await wrapper.setProps({
          modelValue: true,
        });
        await nextTick();
        const ul = wrapper.find("ul");
        await ul.trigger("keydown", { code: Code.UP });
        expect(wrapper.vm.dropdownService.selectPrevious).toHaveBeenCalled();
      });
    });

    it("should close menu and focus trigger on ESC", async () => {
      const slots = {
        trigger: getTrigger(),
        default: [getLi(), getLi()],
      };
      const wrapper = getWrapper({ slots });
      await wrapper.setProps({
        modelValue: true,
      });
      await nextTick();
      const ul = wrapper.find("ul");
      await ul.trigger("keydown", { code: Code.ESCAPE });
      expect(wrapper.vm.localOpen).toBe(false);
    });
  });
  it("should focus first element when opened", async () => {
    const slots = {
      trigger: getTrigger(),
      default: [getLi(), getLi()],
    };
    const wrapper = getWrapper({ slots });
    jest.spyOn(wrapper.vm.dropdownService, "select");
    await nextTick();
    await wrapper.setProps({
      modelValue: true,
    });
    await nextTick();

    expect(wrapper.vm.dropdownService.select).toHaveBeenCalledWith(
      wrapper.vm.dropdownService.items[0]
    );
  });
  it("should close when close event comes from FeatherMenu", async () => {
    const slots = {
      trigger: getTrigger(),
      default: [getLi(), getLi()],
    };
    const wrapper = getWrapper({
      slots,
    });
    await nextTick();
    await wrapper.setProps({
      modelValue: true,
    });
    await nextTick();
    await wrapper.findComponent({ ref: "menu" }).vm.$emit("close");

    expect(wrapper.vm.localOpen).toBe(false);
    expect(getCalls<[boolean]>(wrapper, "update:modelValue")[1][0]).toBe(false);
  });
  it("should have the correct ids labelling the menu", async () => {
    const slots = {
      trigger: getTrigger(),
      default: [getLi(), getLi()],
    };
    const wrapper = getWrapper({ slots });
    await wrapper.setProps({ modelValue: true });
    await nextTick();
    const menu = wrapper.find("ul");
    expect(menu.attributes()["aria-labelledby"]).toBe(
      (wrapper.vm.$refs.menu as { triggerId: string }).triggerId
    );
  });
  describe("a11y", () => {
    it("should be accessible when open", async () => {
      const slots = {
        trigger: getTrigger(),
        default: [getLi(), getLi()],
      };
      const wrapper = getWrapper({ slots });
      await nextTick();

      await wrapper.setProps({ modelValue: true });
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
});
