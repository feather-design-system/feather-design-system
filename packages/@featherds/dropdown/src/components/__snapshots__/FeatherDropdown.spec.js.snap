// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherDropdown.vue should pass through cover property 1`] = `
<div
  class="feather-menu feather-dropdown-container"
>
  
  
  <button
    menu-trigger=""
  >
    TEST
  </button>
  
  
  <teleport-stub>
    <div
      class="feather-menu-dropdown"
      id="feather-menu-dropdown"
      style="width: auto; display: none;"
    >
      <div
        tabindex="0"
      />
      
      <ul
        aria-labelledby="feather-menu-trigger"
        class="feather-dropdown"
      >
        
        <li>
          <a>
            TEST
          </a>
        </li>
        <li>
          <a>
            TEST
          </a>
        </li>
        
      </ul>
      
      <div
        tabindex="0"
      />
    </div>
  </teleport-stub>
</div>
`;

exports[`FeatherDropdown.vue should pass through right property 1`] = `
<div
  class="feather-menu feather-dropdown-container"
>
  
  
  <button
    menu-trigger=""
  >
    TEST
  </button>
  
  
  <teleport-stub>
    <div
      class="feather-menu-dropdown"
      id="feather-menu-dropdown"
      style="width: auto; display: none;"
    >
      <div
        tabindex="0"
      />
      
      <ul
        aria-labelledby="feather-menu-trigger"
        class="feather-dropdown"
      >
        
        <li>
          <a>
            TEST
          </a>
        </li>
        <li>
          <a>
            TEST
          </a>
        </li>
        
      </ul>
      
      <div
        tabindex="0"
      />
    </div>
  </teleport-stub>
</div>
`;
